File: src/com/diffplug/common/rx/RxBox.java
/*
 * Copyright 2016 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.common.rx;

import java.util.function.Consumer;
import java.util.function.Function;

import rx.Observable;

import com.diffplug.common.base.Box;
import com.diffplug.common.base.Consumers;
import com.diffplug.common.base.Converter;

/**
 * {@link RxGetter} and {@link Box} combined in one: a value you can set, get, and subscribe to.
 */
public interface RxBox<T> extends RxGetter<T>, Box<T> {
	/** Returns a read-only version of this `RxBox`. */
	default RxGetter<T> readOnly() {
		return this;
	}

	/** Maps one `RxBox` to another `RxBox`. */
	@Override
	default <R> RxBox<R> map(Converter<T, R> converter) {
		return new RxBoxImp.Mapped<>(this, converter);
	}

	/**
	 * Provides a mechanism for enforcing an invariant.
	 *
	 * For example, let's say that there is an `RxBox<List<T>>` which represents
	 * a user's multiselection in some widget. You would like the user to select
	 * a pair of values - you want to enforce that the selection doesn't grow
	 * beyond 2 elements.
	 *
	 * You could write code like this:
	 *
	 * ```java
	 * RxBox<List<T>> selection = widget.rxSelection();
	 * RxGetter<List<T>> onlyPairs = selection.map(list -> list.stream().limit(2).collect(Collectors.toList()));
	 * Rx.subscribe(onlyPairs, selection);
	 * ```
	 *
	 * But there's a subtle bug in this code.  Because {@link RxGetter} and {@link RxBox} only
	 * fire their observables when their value changes, the selection invariant won't be enforced
	 * unless `onlyPairs` actually changes.  It's easy to fix:
	 *
	 * ```java
	 * RxBox<List<T>> selection = widget.rxSelection();
	 * Observable<List<T>> onlyPairs = selection.asObservable().map(list -> list.stream().limit(2).collect(Collectors.toList()));
	 * Rx.subscribe(onlyPairs, selection);
	 * ```
	 *
	 * But there's still a problem - `selection` will momentarily contain values which violate the invariant. That
	 * means you can't write listener code that relies on the invariant being true.  Which is why the `enforce()` method exists:
	 *
	 * ```java
	 * RxBox<List<T>> onlyPairs = widget.rxSelection().enforce(list -> list.stream().limit(2).collect(Collectors.toList()));
	 * ```
	 *
	 * This creates a new `RxBox` which will always satisfy the invariant, and it will enforce this invariant
	 * on the original `RxBox`. 
	 */
	default RxBox<T> enforce(Function<? super T, ? extends T> enforcer) {
		// this must be a plain-old observable, because it needs to fire
		// every time an invariant is violated, not only when a violation
		// of the invariant causes a change in the output
		Observable<T> mapped = asObservable().map(enforcer::apply);
		Rx.subscribe(mapped, this::set);
		// now we can return the RxBox
		return from(map(enforcer), Consumers.compose(enforcer, this::set));
	}

	/** Creates an `RxBox` with the given initial value. */
	public static <T> RxBox<T> of(T initial) {
		return new RxBoxImp<T>(initial);
	}

	/** Creates an `RxBox` which implements the "getter" part with `RxGetter`, and the setter part with `Consumer`. */
	public static <T> RxBox<T> from(RxGetter<T> getter, Consumer<T> setter) {
		return new RxBox<T>() {
			@Override
			public Observable<T> asObservable() {
				return getter.asObservable();
			}

			@Override
			public T get() {
				return getter.get();
			}

			@Override
			public void set(T value) {
				setter.accept(value);
			}
		};
	}
}

File: src/main/java/com/diffplug/common/rx/RxBox.kt
/*
 * Copyright 2020 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.common.rx

import com.diffplug.common.base.Box
import com.diffplug.common.base.Converter
import com.diffplug.common.rx.Rx.subscribe
import java.util.function.Consumer
import java.util.function.Function
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

/** [RxGetter] and [Box] combined in one: a value you can set, get, and subscribe to. */
interface RxBox<T> : RxGetter<T>, Box<T> {
	/** Returns a read-only version of this `RxBox`. */
	fun readOnly(): RxGetter<T> {
		return this
	}

	/** Maps one `RxBox` to another `RxBox`. */
	override fun <R> map(converter: Converter<T, R>): RxBox<R> {
		return RxBoxImp.Mapped(this, converter)
	}

	/**
	 * Provides a mechanism for enforcing an invariant on an existing `RxBox`.
	 *
	 * The returned `RxBox` and its observable will **always** satisfy the given invariant. If the
	 * underlying `RxBox` changes in a way which does not satisfy the invariant, it will be set so
	 * that it does match the invariant.
	 *
	 * During this process, the underlying `RxBox` will momentarily fail to meet the invariant, and
	 * its `Observable` will emit values which fail the invariant. The returned `RxBox`, however, will
	 * always meet the invariant, so downstream consumers can rely on the invariant holding true at
	 * all times.
	 *
	 * The returned `RxBox` can be mapped, and has the same atomicity guarantees as the underlying
	 * `RxBox` (e.g. an enforced [RxLockBox] can still be modified atomically).
	 *
	 * Conflicting calls to `enforce` can cause an infinite loop, see [Breaker] for a possible
	 * solution.
	 *
	 * ```java
	 * // this will not end well...
	 * RxBox.of(1).enforce(Math::abs).enforce(i -> -Math.abs(i));
	 * ```
	 */
	fun enforce(enforcer: Function<in T, out T>): RxBox<T> {
		// this must be a plain-old observable, because it needs to fire
		// every time an invariant is violated, not only when a violation
		// of the invariant causes a change in the output
		val mapped = asFlow().map { t: T -> enforcer.apply(t) }
		subscribe(mapped) { value: T -> this.set(value) }
		// now we can return the RxBox
		return map(Converter.from(enforcer, enforcer))
	}

	companion object {
		/** Creates an `RxBox` with the given initial value. */
		@JvmStatic
		fun <T> of(initial: T): RxBox<T> {
			return RxBoxImp(initial)
		}

		/**
		 * Creates an `RxBox` which implements the "getter" part with `RxGetter`, and the setter part
		 * with `Consumer`.
		 */
		@JvmStatic
		fun <T> from(getter: RxGetter<T>, setter: Consumer<T>): RxBox<T> {
			return object : RxBox<T> {
				override fun asFlow(): Flow<T> {
					return getter.asFlow()
				}

				override fun get(): T {
					return getter.get()
				}

				override fun set(value: T) {
					setter.accept(value)
				}
			}
		}
	}
}
